lib/concurrent/atomic/semaphore.rb

```diff
@@ -141,12 +145,13 @@ def try_acquire_now(permits)
     end
 
     def try_acquire_timed(permits, timeout)
-      remaining = Condition::Result.new(timeout)
-      while !try_acquire_now(permits) && remaining.can_wait?
-        @condition.signal
-        remaining = @condition.wait(@mutex, remaining.remaining_time)
-      end
-      remaining.can_wait? ? true : false
+      ns_wait_until(timeout) { try_acquire_now(permits) }
```

Comment:
Sorry I am travelling for few days, so I am slow in replies. Yes that's correct. `ns_wait_until` returns `true` only if the condition holds, or `false` on timeout if supplied. It should not return in any other circumstances.